### DISCUSSION TITLE
Queries on Threats Manager Studio for Microsoft Threat modeling tool users

### DISCUSSION THREAD

**Discussion by rsrinivasanhome** [simonec73/threatsmanager#10] [Q&A] [answered]
I am well versed in the Microsoft threat modelling tool and looking at using Threats Manager Studio .  
Can you point me to some documentation or provide information on the following area(s) ?

- What are the main areas that threat manager studio is better than Microsoft threat modelling tool? What extra threats can this detect and why ?
- Do we have any documentation on how to use threats manager studio for user(s) who are accustomed to the Microsoft threat modelling tool ?
- How do we create rules to generate threats?
- How do we represent a http or tcp /ip connection via flow ?
- Till now most of our threat models are in Microsoft threat modelling tool .  Till we get more comfortable with Threats Manager Studio would it add value by generating the threats using threat manager studio and keeping the threat model definition in the Microsoft threat modelling tool ?  
- I observed there are not many policies uploaded so far wanted your opinion on the following 

1. Can we  import a microsoft threat model template (https://github.com/AzureArchitecture/threat-model-templates) into threat manager studio ?
2. Can we generate a template from existing threat model ?

**Comment by simonec73** [comment, marked as answer]
Thank you for your message.

You can find the documentation at https://threatsmanager.com/training/. You can also find five articles on the Home page in the Learn the Threat Modeling Process with TMS section. Here are the direct links for your convenience:
- https://threatsmanager.com/2020/11/21/threat-modeling-with-tms-diagramming/
- https://threatsmanager.com/2020/11/22/threat-modeling-with-tms-threats-and-mitigations/
- https://threatsmanager.com/2020/11/29/threat-modeling-with-tms-adding-your-value-in/
- https://threatsmanager.com/2020/12/06/threat-modeling-with-tms-quality-and-communication/
- https://threatsmanager.com/2020/12/20/threat-modeling-with-tms-the-interviews/

I started out working with TMT. While I've seen various other free tools before working on TMS, my experience was almost exclusively on TMT. For this reason, I've designed it to address what I perceived as the main shortcomings of TMT. Over the years, I've received feedback from dozens of colleagues. As a result, my tool has grown to cover all the various TMT features and improve over the latter's experience. Nowadays, I would say that TMS is better than TMT in every aspect. Of course, I'm its author; therefore, you should not take my view for granted. You should see for yourself and decide if I'm biased or not.

Threat generation and Mitigation association require the Automatic Threat Generation Extension Library, which you can find on the [Downloads](https://threatsmanager.com/downloads/) page. It provides a graphical environment to create generation rules via dragging and dropping. Please refer to https://threatsmanager.com/training/expert/extensions/autothreatgen-rules-definition/ for the details.

For representing flows like an HTTP channel or a TCP connection, you need to use a feature called Item Templates. First of all, to create stencils, you need to work in Expert or Pioneer Execution Mode. If you do not know what it does mean, please refer to https://threatsmanager.com/training/intro/start-here/execution-mode/.  At this point, you would be able to go to the Item Template List in the View ribbon, and then you can create a Flow Template by clicking on the Add Flow Template from the Item Template List ribbon. As a result, you will see a dialog where you should include the details of your new flow type.
Another approach is to create a diagram, select a flow with the same characteristics of the HTTP flow you want to create, and then click Derive Template from its context menu.

Not sure I understand the following question: why would you want to maintain two tools? Do you know that with the MS Threat Modeling Tool Import Extension Library, you can import your .TM7 and .TB7 files? Yes, even templates! Please get a look at https://threatsmanager.com/training/advanced/extensions/tmt-importer/ for more information.
For example, after you import a .TB7 file, you can export a TMS template you can use to create other Threat Models. You can find further details about this process in https://threatsmanager.com/training/expert/templates/templates-creation/. 
And yes, creating a reference Threat Model and then exporting it as a Template is precisely the pattern to develop Templates with TMS.

I hope I have answered all your questions. Please feel free to ask again for any additional doubt.

Thank you for giving TMS a try!

**Reply by rsrinivasanhome** [reply]
I have started to experiment and a few more questions have arisen 

- Has anyone created a AWS template so far ?
- I have been able to add a new Flow (example tcp ip ) but in the GUI how do I select this new flow ?
- How do we track the closure of threat events ?
- Do we track this by the threat event mitigation status ?
- Do we use the directives (when displaying mitigation)  like the justification of microsoft threat modelling tool ?
- What is the functionality of Review Notes used for ? (I tried search in the site but could not get link to the documentation)
- What is the functionality of question list and import questions to be used for ?
- How doe the menu option Generate Topics work ?
- In the following link - https://threatsmanager.com/training/expert/extensions/autothreatgen-rules-definition/ . Where do we click to get the "Is Managed" and where do we get the ignore option ?
Is there any youtube video demo of using this feature ?  That would be very helpful

Does the OR criteria require two options to compare so it can be successfully tested ?  If I  give a single condition it 
seems not to filter .
![image](https://user-images.githubusercontent.com/31912418/137096958-8fa0c867-2e3e-4fa9-b3b2-9a4613cce728.png)
If I remove the condition Object type is Data Store and click on test I do not get any  value in the "Selected External interactors"  text box .  Why could that be ?

When importing a  particular threat model I am getting an error ?  How do I troubleshoot ?
![image](https://user-images.githubusercontent.com/31912418/137117253-14bead39-abf1-4b50-a758-14a74e949121.png)

Standard mitigation -  Does it means a mitigation which needs to be implemented once for the whole system ?
and All threat events linked to a standard mitigation are mitigated.

Non standard mitigation  - A mitigation which needs to be configured or implemented for each threat event ?
All threat events linked to a non standard mitigation may or may not be mitigated

Is the justification field in microsoft threat modeling tool migrated into threats manager studio - threat event ?

@simonec73  based on what I have seen so far the tool looks awesome .

**Comment by simonec73** [comment]
Thank you for your feedback and for the questions!

I will try to answer, here:
- I do not know, sorry, but I guess not. The tool has been published about 11 months ago, and I haven't seen any template produced by the Community, so far. Unfortunately, I have no time to work on producing new templates. It would be great if the Community starts producing those templates. In any case, I am working on a couple of new functionalities that should simplify the creation of templates, and hopefully, even reduce the need.
- If you have created a new Flow as an Item Type, to create it in the Diagram you have first to create a normal Flow, then to open the context menu and select Change Template. At this point, you can select the desired type of Flow.
- Threat Events have a Severity, which represents the current evaluation, that is the current residual risk. To mark a Threat Event as addressed, you have to change its Severity to Info.
- Threat Event Mitigation Status is used to track the status of the assigned Mitigation. Of course, there is a relationship between the two: in general, the residual risk equals the original risk minus the combined effect of the Implemented and Existing mitigations. The next version will introduce functionalities to assist this evaluation. Essentially, you will be able to see when there is a discrepancy between the expected severity and the current one and will give you the opportunity to fix it or to introduce compensations to account for the difference: for example, the risk may be higher than usual because the processed data is particularly sensitive.
- No: directives are used to provide additional special indications. For example, you may have a mitigation stating that you need to validate the input. The directive would specify which type of validations to apply, or what libraries to use. The main equivalent of the Justification is represented by the Mitigations themselves, at least for those cases where the status of the Threat in TMT was Mitigated.
- Review Notes is a new functionality introduced with v1.4.0. I confess I've been a little lazy in updating the documentation. Hopefully, I will address this after the upcoming version 1.5.0. But you need to know that this is a passion project, done in my spare time, on top of my daily job, stealing time from my family and from other activities I need to preserve my own sanity. This is the reason why it is of utter importance for this initiative to succeed, that the Community of Security users starts contributing to it, for example providing Templates to cover the need. Of course, I can provide the guidance required.
- Questions are yet another capability introduced with version 1.4.x. The idea is to allow to annotate any object, like Entities, Flows, etc. Questions build upon the annotation capabilities, providing a way to define questions with associated Automatic Generation Rules. So, you import those questions, then you click a button and... bam! Your Threat Model is annotated with Topics to be Clarified. Each Topic is generated by a Question and represents a piece of information you should gather to get a better understanding of the solution. This allows also to track the answers. Furthermore, the Annotations button in the Main ribbon allows to search all the annotations in the Threat Model, and even to export all unanswered Topics as a CSV or an Excel file. The idea is that those files could be sent to the experts of the solution, to have their answers. If you use something like OneDrive for Business, you would even be able to get their answers in real-time! I can see the possibility to implement further capabilities in the future, to extend this and create a full lifecycle, where you could publish your additional questions in the same Excel file, without disturbing what is already there, and you could even retrieve the answers and import them in the Threat Model. How that sounds? :)
- Generate topics is the button that generates the Topics to be Clarified out of the Questions, using the Automatic Generation Rules.
- Is Managed and Ignore are custom properties defined in the Threat Model I used as an example. Other Threat Models may have other properties defined. The Auto Threat Gen dialog automatically picks the custom properties defined in the Threat Model. I agree for the video. This definitely is something I want to include.
- OR is an N-ary operator. As such, requires two or more values to compare. If you include only one, it would not work. If you want only one, you can simply drag and drop it on the empty space. If you want to clear the tree, you can use Clear from the context menu. The problem you discuss under the picture is due to this reason.
- Next version introduces some fixes to the Threat Model import function. A reason may be that the file is malformed. Can you open it with TMT? My telemetry shows this error, repeated 5 times about 8 hours ago: "The input is not a valid Base-64 string as it contains a non-base 64 character, more than two padding characters, or an illegal character among the padding characters". If it is your attempt to load the file that generated the error and you can open it with TMT, then I may need to fix my code to account for this situation.
- Standard Mitigations: a standard mitigation is a mitigation already associated with the Threat Type. As such, it is known to be a good answer for that Threat Type. For more information, please refer to https://threatsmanager.com/training/intro/basic/mitigations/.
- Non-Standard Mitigations are simply known mitigations which are not yet been associated with the Threat Type.
- The justification field is migrated as a new property, along with all the other properties, when you import a TM7 file.

**Reply by rsrinivasanhome** [reply]
I totally agree with you. Templates is something the community should develop.
The threat model I had imported was a valid one so I think the code needs to be fixed.
Thank you for time and effort you have spent in answering the questions .  It has been very informative.

**Reply by simonec73** [reply]
Thank you. The problem is that the template used to create the document has a misconfigured image for one or more stencils. I am not sure how that could happen. My tool complains because the image is not Base64 encoded. I've already fixed it within the upcoming Version 1.5.0. The new behavior is that entities with an incorrect image are provided with the default icon from Threats Manager Studio.

**Reply by rsrinivasanhome** [reply]
Reference:-  https://threatsmanager.com/training/advanced/threats-and-mitigations/roadmap/

"Another effect of the configuration of the Residual Risk Estimator is that the Mitigations in the upper section of the Roadmap tool are colored based on how effective they are in reducing the overall Risk: the most effective third of the Mitigations gets a dark green background and white text; the second third of the Mitigations gets a light green background and black text; and finally the least effective third gets a white background and black text. This allows to easily discriminate the Mitigations you should focus on."

- What is the criteria to detect the effectiveness of a mitigation ?Is this based on mitigation status and strength ?   Is not the residual risk estimator used to plot the acceptable risk line ?

- In the roadmap feature we do not consider threats without mitigation am I correct?

- In the bar chart how are the bar heights calculated . Does it depend on the mitigation status ?

In the roadmap display if we plan all the mitigations to be completed in mid term will there be any bar for long term in the bar chart ?
![image](https://user-images.githubusercontent.com/31912418/137747539-8fcff866-883c-48cf-84df-ba61d7ce119e.png)

**Reply by simonec73** [reply]
Hi, those are my answers:
- effectiveness is calculated considering the Threat events interested by the mitigation, their current severity and the strength of the mitigation. For example, a mitigation which affects a single high severity mitigation with an average strength will be considered less effective than a mitigation affecting 4 Medium severity threats, when they are considered strong. The math is quite simple, really.
- In the roadmap we consider only mitigations, therefore yes it's correct.
- The basic idea is discussed  on
https://simoneonsecurity.com/2020/04/02/project-managers-and-threat-modeling-vnext/.
- Yes, it will have the same length than the mid term column.

Thank you again for your questions. 

For the future, to simplify retrieving info from our discussion, I would suggest to create a new conversation for each question you have.
